src/dbnode/storage/index.go

```diff
@@ -716,77 +723,196 @@ func (i *nsIndex) Query(
+			wg.Add(1)
+			running := workers.GoWithTimeout(func() {
+				execBlockQuery()
+				wg.Done()
+			}, timeLeft)
+			timedOut = !running
+		} else {
+			timedOut = true
+		}
+
+		if timedOut {
+			// Exceeded our deadline waiting for this block's query to start
+			return index.QueryResults{},
+				fmt.Errorf("index query exceeded timeout: %s", timeout.String())
+		}
+	}
+
+	// Wait for queries to finish
+	wg.Wait()
+
+	// NB(r): Dont' need to hold lock for concurrent results since
+	// at this point all block queries must have finished.
+	// If we ever to early aborts we'll have to make sure that the results
+	// aren't modified by aborted goroutines we launched that still have
+	// access to the concurrent results set (which is difficult because
+	// the reader of the results will also need to hold lock instead of just
+	// being returned unsynchronized results from this method like we
+	// currently do).
+	if err := multiErr.FinalError(); err != nil {
+		return index.QueryResults{}, err
+	}
+
+	return index.QueryResults{
+		Exhaustive: exhaustive,
+		Results:    results.Results(),
+	}, nil
+}
+
+func (i *nsIndex) timeoutForQueryWithLock(
+	ctx context.Context,
+) time.Duration {
+	// TODO(r): Allow individual queries to specify timeouts using
+	// deadlines passed by the context
+	return i.state.runtimeOpts.defaultQueryTimeout
+}
+
+func (i *nsIndex) overriddenOptsForQueryWithLock(
+	opts index.QueryOptions,
+) index.QueryOptions {
+	// Override query response limit if needed
 	if i.state.runtimeOpts.maxQueryLimit > 0 && (opts.Limit == 0 ||
 		int64(opts.Limit) > i.state.runtimeOpts.maxQueryLimit) {
 		i.logger.Debugf("overriding query response limit, requested: %d, max-allowed: %d",
 			opts.Limit, i.state.runtimeOpts.maxQueryLimit) // FOLLOWUP(prateek): log query too once it's serializable.
 		opts.Limit = int(i.state.runtimeOpts.maxQueryLimit)
 	}
+	return opts
+}
 
-	var (
-		exhaustive = true
-		results    = i.opts.IndexOptions().ResultsPool().Get()
-		err        error
-	)
-	results.Reset(i.nsMetadata.ID())
-	ctx.RegisterFinalizer(results)
-
+func (i *nsIndex) blocksForQueryWithLock(queryRange xtime.Ranges) ([]index.Block, error) {
 	// Chunk the query request into bounds based on applicable blocks and
 	// execute the requests to each of them; and merge results.
-	queryRange := xtime.NewRanges(xtime.Range{
-		Start: opts.StartInclusive, End: opts.EndExclusive})
+	blocks := make([]index.Block, 0, len(i.state.blockStartsDescOrder))
 
 	// iterate known blocks in a defined order of time (newest first) to enforce
 	// some determinism about the results returned.
 	for _, start := range i.state.blockStartsDescOrder {
 		block, ok := i.state.blocksByTime[start]
 		if !ok { // should never happen
-			return index.QueryResults{}, i.missingBlockInvariantError(start)
+			return nil, i.missingBlockInvariantError(start)
 		}
 
-		// ensure the block has data requested by the query
+		if queryRange.IsEmpty() {
```

Comment:
nit: This might be a little less confusing if it was the first part of the loop. As is, I was confused why you would need to check this after doing the block lookup, but its not related to that at all